Add iframe depth limit

nt1m opened the pull request:
#### 65071a674a05a195850647ffccfb78b36fc22000
<pre>
Add iframe depth limit
<a href="https://bugs.webkit.org/show_bug.cgi?id=67940">https://bugs.webkit.org/show_bug.cgi?id=67940</a>
rdar://101560112

Reviewed by Darin Adler and Alex Christensen.

* LayoutTests/fast/frames/frame-depth-limit-expected.txt: Added.
* LayoutTests/fast/frames/frame-depth-limit.html: Added.
* LayoutTests/fast/frames/resources/self-referential-iframe.html: Added.
* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::loadSubframe):
* Source/WebCore/page/FrameTree.cpp:
(WebCore::FrameTree::depth const):
* Source/WebCore/page/FrameTree.h:
* Source/WebCore/page/Page.h:

Canonical link: <a href="https://commits.webkit.org/257550@main">https://commits.webkit.org/257550@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd2d9d308c59c04f90ce137f566dcdf9b7cdfc27

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99248 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8446 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32367 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108632 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168880 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103243 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85766 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91741 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106564 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105004 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6794 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90351 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33796 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88614 "Found 1 new API test failure: TestWebKitAPI.WebKit.QuotaDelegateReload (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21698 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2326 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23214 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2221 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5192 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/7466 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42688 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3696 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/Page.h:
```diff
@@ -321,6 +321,7 @@ class Page : public Supplementable<Page>, public CanMakeWeakPtr<Page> {
     WEBCORE_EXPORT static unsigned nonUtilityPageCount();
 
     unsigned subframeCount() const;
+    unsigned subframeDepth() const;
```
Unused

darinadler commented on Source/WebCore/page/FrameTree.h:
```diff
@@ -75,13 +75,16 @@ class FrameTree {
 
     void resetFrameIdentifiers() { m_frameIDGenerator = 0; }
 
+    static const int maxFrameDepth = 32;
```
`constexpr`

Not sure this class is the best place for this limit, but I don’t have a strong opinion about a better place.

achristensen07 replied:
I think this and Page::maxNumberOfFrames should be in the same place.

darinadler commented on LayoutTests/fast/frames/frame-depth-limit.html:
```diff
@@ -0,0 +1,39 @@
+            testRunner.notifyDone();
+        };
+
+        window.addEventListener("load", () => {
```
No need for `window.` here.

darinadler commented on LayoutTests/fast/frames/frame-depth-limit.html:
```diff
@@ -0,0 +1,39 @@
+
+        window.addEventListener("load", () => {
+            let delay = setTimeout(finishTest, 500);
+            window.addEventListener("message", () => {
```
No need for `window.` here.

darinadler commented on LayoutTests/fast/frames/frame-depth-limit.html:
```diff
@@ -0,0 +1,39 @@
+        }
+
+        let finishTest = () => {
+            testRunner.notifyDone();
```
A little strange that we check for `window.testRunner` in the code running at the start of the test, but not at the finish. I suggest having the check here too.

Could use DOM mutation to add a message here so we would not have to deduce from lack of hang that we got done. Especially useful if you run this in browser instead of with `testRunner`.

darinadler commented on LayoutTests/fast/frames/resources/self-referential-iframe.html:
```diff
@@ -0,0 +1,35 @@
+    <script>
+        window.addEventListener('load', () => {
+            let iframe = document.getElementsByTagName('iframe')[0];
+            let number = Math.random();
```
There’s no strong guarantee we won’t have collisions with `random`. Is there another way to do this?

achristensen07 replied:
A better way to do this would be to increment the number from this frame for the next frame.  We could then also use this to do something like verify that load fails exactly where we want it to, then postMessage a success message back up the frame tree.

darinadler commented on LayoutTests/fast/frames/resources/self-referential-iframe.html:
```diff
@@ -0,0 +1,35 @@
+            let iframe = document.getElementsByTagName('iframe')[0];
+            let number = Math.random();
+            iframe.src = `self-referential-iframe.html?foo=${number}`;
+            window.top.postMessage(number, "*");
```
No need for `window.` here.

darinadler commented on LayoutTests/fast/frames/resources/self-referential-iframe.html:
```diff
@@ -0,0 +1,35 @@
+        }
+    </style>
+    <script>
+        window.addEventListener('load', () => {
```
No need for `window.` here.

darinadler commented on LayoutTests/fast/frames/resources/self-referential-iframe.html:
```diff
@@ -0,0 +1,35 @@
+    </script>
+</head>
+<body>
+    <div class="container">
+    <div class="container">
+    <div class="container">
```
Why the three nested containers?

nt1m replied:
This is just to trigger the creation of layers to stress test the stack.

achristensen07 commented on LayoutTests/fast/frames/resources/self-referential-iframe.html:
```diff
@@ -0,0 +1,35 @@
+    <div class="container">
+    <div class="container">
+        <iframe scrolling="no"></iframe>
+    </div>
```
nit: don't we need 3 of these?

achristensen07 commented on Source/WebCore/page/FrameTree.cpp:
```diff
@@ -491,6 +491,14 @@ AbstractFrame& FrameTree::top() const
+{
+    unsigned depth = 0;
+    for (auto* parent = &m_thisFrame; parent; parent = parent->tree().parent())
+        depth++;
```
We could also write a function isDepthGreaterThanOrEqualTo that returns early, but since we are using this to limit the depth, I think they're basically equivalent in this case.

nt1m committed:
Add iframe depth limit
https://bugs.webkit.org/show_bug.cgi?id=67940
rdar://101560112

Reviewed by Darin Adler and Alex Christensen.

* LayoutTests/fast/frames/frame-depth-limit-expected.txt: Added.
* LayoutTests/fast/frames/frame-depth-limit.html: Added.
* LayoutTests/fast/frames/resources/self-referential-iframe.html: Added.
* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::loadSubframe):
* Source/WebCore/page/FrameTree.cpp:
(WebCore::FrameTree::depth const):
* Source/WebCore/page/FrameTree.h:
* Source/WebCore/page/Page.h:

Canonical link: https://commits.webkit.org/257550@main
